### ISSUE TITLE
503s when bringing up pods in 1.8

### ISSUE THREAD

**Issue by lonnix** [istio/istio#30114]
**Bug description**
I have an app running with the istio-proxy container being injected into it. When I do a deployment or scale up the number of pods I notice a brief period of 503s, maybe 5-10s. This happens consistently. Without the istio-proxy sidecar this issue does not happen. We have liveness and readiness probes in place with an initial delay so our pods do not register as ready until the app has fully spun up. However with istio-proxy it seems like the pod starts taking traffic before it is actually ready. The app has its own init container, so there are 4 containers in the pod: istio-proxy, istio-init, init-container, app.

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
The creation of new pods should not cause 503 errors when running with the istio-proxy sidecar

**Steps to reproduce the bug**
1) deploy an app with the istio-proxy sidecar that takes at least 5 seconds to start up
2) create load on this app
3) scale up the number of pods
4) observe a brief burst of 503 response codes

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
Istio version is 1.8.0
```
~: helm version --short
v3.4.1+gc4e7485
~: kubectl version --short
Client Version: v1.18.5
```

**How was Istio installed?**
Helm

**Environment where the bug was observed (cloud vendor, OS, etc)**
All of my clusters run on EKS, kubernetes version 1.17

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by hzxuzhonghu** [comment]
Have you tried `holdApplicationUntilProxyStarts` https://github.com/istio/api/pull/1670

**Comment by lonnix** [comment]
@hzxuzhonghu Yes, I gave that a try and it didn't help. I almost need the opposite of that, something like `holdProxyUntilApplicationStarts` since my issue seems to be the proxy being ready too soon

**Comment by howardjohn** [comment]
> We have liveness and readiness probes in place with an initial delay so our pods do not register as ready until the app has fully spun up. However with istio-proxy it seems like the pod starts taking traffic before it is actually ready.

This should not happen. Kubernetes will mark the pod ready when all containers are ready. Therefor, if your container is not passing readiness probe, for say 10s, then the pod should NOT get traffic.

Are you certain it is getting traffic **before** your pod passes the readiness probe?

**Comment by rnrsr** [comment]
> Are you certain it is getting traffic before your pod passes the readiness probe?

we have exactly the same problem, and the answer is yes, as soon as we run a Helm deployment the service becomes unavailable with 503s. Even when 0/3 containers are ready and they are still Initializing. It happens almost immediately after the rollout begins, and lasts until the first (new) pod is available with all containers ready.

**Comment by howardjohn** [comment]
oh so its not new pods coming up, its the old ones going down? Have you taken a look at https://deepsource.io/blog/zero-downtime-deployment/? this would happen without Istio as well - rollouts are *not* zero downtime if your application exits immediately on SIGTERM

**Comment by rnrsr** [comment]
> oh so its not new pods coming up, its the old ones going down?

No it happens when the new pods appear, we see this immediately after the deploy. The old pods are not terminated.

Updating our istio planes to 1.7 and adding `holdApplicationUntilProxyStarts` to the istioctl install seems to have fixed it, in our case.

**Comment by rnrsr** [comment]
Actually, i believe you are right. The confusion comes from seeing new pods appear and _then_ seeing the 503s. But if we monitor the `endpoints` in kubectl we see the _old_ endpoint hang around long enough for it to outlive the old pods, which are, as you predicted, being killed very abruptly. 

We thought 1.7 had fixed it because we saw the first few deploys work without triggering a 503, but then it started happening again. Turns out we need `preStop` hooks. Adding them and making sure the deployment jobs were not selected by the service selector solved the problem. Thanks for the link @howardjohn !
